security/pkg/nodeagent/caclient/providers/spire/client.go

```diff
@@ -0,0 +1,375 @@
+	if err != nil {
+		log.Errorf("Failed opening stream to attestating SPIRE server: %v", err)
+		return nil, errors.New("failed opening stream to attestating SPIRE server")
+	}
+
+	if err := attestStream.Send(attestReq); err != nil {
+		log.Errorf("Not able to send attestation request to SPIRE server: %v", err)
+		return nil, fmt.Errorf("not able to send attestation request to SPIRE server")
+	}
+
+	attestResp, err := attestStream.Recv()
+	if err != nil {
+		return nil, fmt.Errorf("attesting to SPIRE server: %v", err)
+	}
+
+	if len(attestResp.SvidUpdate.Svids) == 0 {
+		return nil, fmt.Errorf("no svid returned from attestation call to SPIRE server")
+	}
+
+	svid, ok := attestResp.SvidUpdate.Svids[certID.String()]
+	if !ok {
+		log.Errorf("SPIRE response does not contains a SVID for %v", certID)
+		return nil, fmt.Errorf("SPIRE response does not contains a SVID for %v", certID)
+	}
+
+	// svid cert chain can contains multiple certificates
+	certificates, err := x509.ParseCertificates(svid.CertChain)
+	if err != nil {
+		log.Errorf("Failed to parse svid certificates from SPIRE: %v", err)
+		return nil, fmt.Errorf("failed to parse svid from SPIRE")
+	}
+
+	// Get bundle for trust domain
+	bundle, ok := attestResp.SvidUpdate.Bundles[fmt.Sprintf(spiffeIDTemplate, certID.Host)]
+	if !ok {
+		log.Errorf("No bundle returned from SPIRE")
+		return nil, fmt.Errorf("no bundle returned from SPIRE")
+	}
+
+	var bundles []*x509.Certificate
+	for _, ca := range bundle.RootCas {
+		c, err := x509.ParseCertificate(ca.DerBytes)
+		if err != nil {
+			log.Error("not able to parse certificate from SPIRE bundle response")
+		} else {
+			bundles = append(bundles, c)
+		}
+
+	}
+
+	certChain, err := c.buildCertChain(certificates, bundles)
+	if err != nil {
+		return nil, err
+	}
+
+	return certChain, nil
+}
+
+// buildCertChain creates a certificate chain based on svid response
+func (c *spireClient) buildCertChain(certificates, bundles []*x509.Certificate) ([]string, error) {
+	certsMap := make(map[string]bool)
+	var certChain []string
+
+	// add certificates
+	for _, cert := range certificates {
+		certPEM := certChainPEM(cert.Raw)
+
+		certsMap[certPEM] = true
+		certChain = append(certChain, certPEM)
+	}
+
+	var latestBundle string
+	var rootCerts string
+
+	r, err := buildFullChain(certificates, bundles)
+	if err != nil {
+		log.Errorf("Failed to build certificate chain from SPIRE server: %v", err)
+		return nil, errors.New("failed to build certificate chain from SPIRE server")
+	}
+
+	// Istio considers the last element in the returned chain the “root” certificate that is used to validate peer certificates.
+	// Lump all of the certificates in the chain of trust that belong to the trust bundle returned by SPIRE into a single entry to be used as the “root”.
+	for _, cert := range r {
+		caPEM := certChainPEM(cert.Raw)
+		if !certsMap[caPEM] {
+			rootCerts += caPEM
+		}
+	}
+	certChain = append(certChain, rootCerts)
+
+	for _, ca := range bundles {
```

Comment:
When enumerating the certificates in bundles, it seems that the order of the certificates matters because the concatenated certificates are used in setLatestBundle() to detect whether the latest trust bundle has changed or not.
E.g., if the certificates in the bundles variable is [cert2, cert1] whereas the latest trust bundle is [cert1, cert2], setLatestBundle() will regard the latest trust bundle has changed. But in this case, the certificates in the trust bundles are the same.

Reply:
I pushed a solution